File: sber_pay_android/android/src/main/kotlin/plugin/sdk/SberPayPlugin.kt
package plugin.sdk

import FlutterError
import InitConfig
import PayConfig
import SberPayApi
import SberPayApiEnv
import SberPayApiPaymentStatus
import android.app.Activity
import android.content.Context
import android.os.Build
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import spay.sdk.SPaySdkApp
import spay.sdk.api.InitializationResult
import spay.sdk.api.PaymentResult
import spay.sdk.api.SPaySdkInitConfig
import spay.sdk.api.SPayStage

/**
 * Плагин для оплаты с использованием SberPay. Для работы нужен установленный Сбербанк (либо Сбол).
 */
class SberPayPlugin : FlutterPlugin, ActivityAware, SberPayApi {

    private lateinit var activity: Activity
    private lateinit var context: Context
    private lateinit var sberPayInstance: SPaySdkApp

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        sberPayInstance = SPaySdkApp.getInstance()
        SberPayApi.setUp(flutterPluginBinding.binaryMessenger, this)
    }

    /**
     * Метод инициализации, выполняется перед стартом приложения.
     *
     * @property InitConfig конфигурация инициализации
     */
    override fun initSberPay(config: InitConfig, callback: (Result<Boolean>) -> Unit) {
        onApiVersionCheck(callback = callback) {
            val sPayStage = when (config.env) {
                SberPayApiEnv.SANDBOX_REAL_BANK_APP -> SPayStage.SandboxRealBankApp
                SberPayApiEnv.SANDBOX_WITHOUT_BANK_APP -> SPayStage.SandBoxWithoutBankApp
                else -> SPayStage.Prod
            }
            val enableBnpl = config.enableBnpl ?: false

            try {
                val sPayConfig = SPaySdkInitConfig(
                    application = activity.application,
                    stage = sPayStage,
                    enableBnpl = enableBnpl
                ) { result: InitializationResult ->
                    when (result) {
                        is InitializationResult.ConfigError -> callback(Result.success(false))
                        InitializationResult.Success -> callback(Result.success(true))
                    }
                }
                sberPayInstance.initialize(sPayConfig)
            } catch (e: Exception) {
                callback(Result.failure(FlutterError("-", e.localizedMessage, e.message)))
            }
        }


    }

    /**
     * Метод для проверки готовности к оплате.
     *
     * @return Если у пользователя нет установленного сбера в режимах
     * SPayStage.SandboxRealBankApp, SPayStage.prod - вернет false.
     */
    override fun isReadyForSPaySdk(): Boolean {
        return onApiVersionCheck { return@onApiVersionCheck sberPayInstance.isReadyForSPaySdk(context) }
    }

    /**
     * Метод для оплаты.
     *
     * @property PayConfig конфигурация оплаты
     * @return SberPayApiPaymentStatus статус оплаты
     */
    override fun payWithBankInvoiceId(config: PayConfig, callback: (Result<SberPayApiPaymentStatus>) -> Unit) {
        onApiVersionCheck(callback) {
            var hasEventSent = false // Флаг для отслеживания отправки события

            try {
                val appPackage = context.packageName
                // Пока поддержка только русского языка
                val language = "RU"

                sberPayInstance.payWithBankInvoiceId(
                    activity,
                    config.apiKey,
                    config.merchantLogin,
                    config.bankInvoiceId,
                    config.bankInvoiceId,
                    appPackage,
                    language
                ) { response ->
                    if (!hasEventSent) {
                        when (response) {
                            // Оплата не завершена
                            is PaymentResult.Processing -> callback(Result.success(SberPayApiPaymentStatus.PROCESSING))
                            // Оплата прошла успешно
                            is PaymentResult.Success -> callback(Result.success(SberPayApiPaymentStatus.SUCCESS))
                            // Оплата отменена
                            is PaymentResult.Cancel -> callback(Result.success(SberPayApiPaymentStatus.CANCEL))
                            // Оплата прошла с ошибкой
                            is PaymentResult.Error -> callback(
                                Result.failure(
                                    FlutterError(
                                        "-", "MerchantError", response.merchantError?.description
                                            ?: "Ошибка выполнения оплаты"
                                    )
                                )
                            )
                        }
                    }
                    hasEventSent = true
                }
            } catch (error: Exception) {
                callback(Result.failure(FlutterError("-", error.localizedMessage, error.message)))
            }
        }
    }

    private fun <T> onApiVersionCheck(callback: (Result<T>) -> Unit, successCallback: () -> Unit) {
        // Минимальная версия Android для SberPay SDK - 24
        if (Build.VERSION.SDK_INT < 24) {
            callback(Result.failure(FlutterError("NOT_SUPPORTED", "Android API Error", "Your app SDK < 24")))

            return
        }

        successCallback()
    }

    private fun <T> onApiVersionCheck(successCallback: () -> T): T {
        // Минимальная версия Android для SberPay SDK - 24
        if (Build.VERSION.SDK_INT < 24) {
            throw FlutterError("NOT_SUPPORTED", "Android API Error", "Your app SDK < 24")
        }

        return successCallback()
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        SberPayApi.setUp(binding.binaryMessenger, null)
    }

    override fun onAttachedToActivity(activityBinding: ActivityPluginBinding) {
        activity = activityBinding.activity
    }

    override fun onReattachedToActivityForConfigChanges(activityBinding: ActivityPluginBinding) {
        activity = activityBinding.activity
    }

    override fun onDetachedFromActivity() {}

    override fun onDetachedFromActivityForConfigChanges() {}
}

File: sber_pay_ios/ios/Classes/SberPayPlugin.swift
import Flutter
import UIKit
import SPaySdk

// This extension of Error is required to do use FlutterError in any Swift code.
extension FlutterError: Error {}

/**
 * Плагин для оплаты с использованием SberPay. Для работы нужен установленный Сбербанк (либо Сбол).
 */
public class SberPayPlugin: NSObject, FlutterPlugin, SberPayApi{
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger : FlutterBinaryMessenger = registrar.messenger()
        let instance = SberPayPlugin()
        SberPayApiSetup.setUp(binaryMessenger: messenger, api: instance)
        
        /// Создание [addApplicationDelegate] для перехода по диплинку обратно в приложение
        registrar.addApplicationDelegate(instance)
    }
    
    public func application(_ app: UIApplication,open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
        /// Если при открытии приложения с диплинком если он содержит хост "spay", то такой диплинк
        /// попадает в нативный плагин. Таким образом работает возврат в приложение и получение данных нативным
        /// SDK от приложения Сбербанк онлайн/СБОЛ
        if  url.host == "spay" {
            SPay.getAuthURL(url)
        }
        
        return true
    }
    
    /**
     Метод инициализации, выполняется перед стартом приложения.
     - Parameter InitConfig конфигурация инициализации
     */
    internal func initSberPay(config: InitConfig, completion: @escaping (Result<Bool, any Error>) -> Void) {
        onApiVersiobCheck(onError: completion) {
            let env = config.env
            let enableBnpl = config.enableBnpl ?? false
            
            let sPayStage: SEnvironment
            switch env {
            case SberPayApiEnv.sandboxRealBankApp:
                sPayStage = .sandboxRealBankApp
            case SberPayApiEnv.sandboxWithoutBankApp:
                sPayStage = .sandboxWithoutBankApp
            default:
                sPayStage = .prod
            }
            
            SPay.setup(bnplPlan: enableBnpl, environment: sPayStage) { error in
                if let error {
                    completion(.failure(FlutterError(code: "-", message: "Ошибка инициализации", details: error.description)))
                } else {
                    completion(.success(true))
                }
            }
        }
    }
    
    /**
     Метод для проверки готовности к оплате.
     
     - Returns Если у пользователя нет установленного сбера в режимах SEnvironment.sandboxRealBankApp,
     SEnvironment.prod - вернет false.
     */
    internal func isReadyForSPaySdk() throws -> Bool {
        return try onApiVersiobCheck() {
            return SPay.isReadyForSPay
        }
    }
    
    /**
     Метод для оплаты.
     - Parameter PayConfig конфигурация оплаты
     - Returns SberPayApiPaymentStatus статус оплаты
     */
    internal func payWithBankInvoiceId(config: PayConfig, completion: @escaping (Result<SberPayApiPaymentStatus, Error>) -> Void) {
        onApiVersiobCheck(onError: completion) {
            if config.bankInvoiceId.count != 32 {
                completion(.failure(FlutterError(code: "-", message: "MerchantError", details: "Длина bankInvoiceId должна быть 32 символа")))
            }
            
            guard let topController = self.getTopViewController() else {
                completion(.failure(FlutterError(code: "PluginError", message: "SberPay: Failed to implement controller", details: nil)))
                return
            }
            
            let request = SBankInvoicePaymentRequest(
                merchantLogin: config.merchantLogin,
                bankInvoiceId: config.bankInvoiceId,
                orderNumber: config.orderNumber,
                language: "RU",
                redirectUri: config.redirectUri,
                apiKey: config.apiKey)
            
            SPay.payWithBankInvoiceId(with: topController, paymentRequest: request) { state, info, localSessionId  in
                switch state {
                case .success:
                    completion(.success(SberPayApiPaymentStatus.success))
                case .waiting:
                    completion(.success(SberPayApiPaymentStatus.processing))
                case .cancel:
                    completion(.success(SberPayApiPaymentStatus.cancel))
                case .error:
                    completion(.failure(FlutterError(code: "-", message: "Ошибка оплаты", details: info)))
                @unknown default:
                    completion(.failure(FlutterError(code: "-", message: "Неопределенная ошибка", details: info)))
                }
            }
        }
    }
    
    private func getTopViewController() -> UIViewController? {
        var topController = UIWindow.window?.rootViewController
        
        while let presentedViewController = topController?.presentedViewController {
            topController = presentedViewController
        }
        
        return topController
    }
    
    private func onApiVersiobCheck<T>(onSuccess: @escaping () -> T)  throws -> T {
        if #unavailable(iOS 13.0) {
            throw FlutterError(code: "NOT_SUPPORTED", message: "iOS API Error", details: "Your iOS version < 13")
        }
        
        return onSuccess()
    }
    
    private func onApiVersiobCheck<T>(onError: @escaping (Result<T, Error>) -> Void, onSuccess: @escaping () -> Void) {
        if #unavailable(iOS 13.0) {
            onError(.failure(FlutterError(code: "NOT_SUPPORTED", message: "iOS API Error", details: "Your iOS version < 13")))
            
            return
        }
        
        onSuccess()
    }
}

extension UIWindow {
    static var window: UIWindow? {
        if #available(iOS 13.0, *) {
            return UIApplication.shared.windows.first { $0.isKeyWindow }
        } else {
            return UIApplication.shared.keyWindow
        }
    }
}
